src/mod_mam_utils.erl

```diff
@@ -234,7 +249,12 @@ maybe_external_binary_to_mess_id(BExtMessID) ->
 %% @doc Decode a message ID received from the user.
 -spec external_binary_to_mess_id(binary()) -> integer().
```

Comment:
spec violation.

Reply:
Ok, sorry, this is a mess. Sorry this is taking so long. Out of curiosity how did you find that? I've had a failing spec reported before but not this time. What did I overlook?

Anyway, I think I need a bit more time again to get this right. I'll introduce a type but want to think a bit to provide a clean solution this time.

Reply:
Dialyzer does not like to find errors in `try...catch`-es.

It's pretty easy, I really dislike functions that return more than one type. ;)
I.e. `integer() | binary()`. Because it's leads to code branching. This means potentially some bugs, specs are the easiest ones.
The next kind of bugs you would need to look at runtime is with some backend modules, expecting integer-id (ok, we have handled this by specifying the case in docs).

Reply:
1. UIDs may be tagged values and backends could report and error when they get an unsupported UID format.
2. Backend may check the configured UID provider on startup and refuse to start when it's not supported.
3. Backends may expose extra `supported_uid_formats/0` callback and `mod_mam` should verify it on startup.
4. UID providers may expose `uid_format/0` callback because there may be several providers that return a valid integer, a string etc., so the validation won't be based only on provider name.

For sake of simplicity we may stick to (3) for now and perhaps extend it with (4) if the UID provider list grows.